vars/runATH.groovy

```diff
@@ -0,0 +1,221 @@
+        List<String> env = [
+                "JAVA_HOME=${tool 'jdk8'}",
+                'PATH+JAVA=${JAVA_HOME}/bin',
+                "PATH+MAVEN=${tool 'mvn'}/bin"
+
+        ]
+
+        if (!fileExists(metadataFile)) {
+            echo "Skipping ATH execution because the metadata file does not exist. Current value is ${metadataFile}."
+            skipExecution = true
+            return
+        }
+
+        stage("Getting ATH sources and Jenkins war") {
+            // Start validation
+            metadata = readYaml(file: metadataFile).ath
+            if (metadata == null) {
+                error "The provided metadata file seems invalid as it does not contain a valid ath section"
+            }
+            if (metadata == 'default') {
+                echo "Using default configuration for ATH"
+                metadata = [:]
+            } else if (metadata.browsers == null) {
+                echo "The provided metadata file does not include the browsers property, using firefox as default"
+            }
+            // Allow to override athUrl and athRevision from metadata file
+            athUrl = metadata.athUrl ?: athUrl
+            isLocalATH = athUrl.startsWith("file://")
+            athRevision = metadata.athRevision ?: athRevision
+
+            // Allow override of jenkins version from metadata file
+            jenkins = metadata.jenkins ?: jenkins
+            isVersionNumber = (jenkins =~ /^\d+([.]\d+)*$/).matches()
+
+            if (!isLocalATH) {
+                echo 'Checking connectivity to ATH sources…'
+                sh "git ls-remote --exit-code -h ${athUrl}"
+            }
+            if (jenkins == "latest") {
+                jenkinsURl = mirror + "war/latest/jenkins.war"
+            } else if (jenkins == "latest-rc") {
+                jenkinsURl = mirror + "/war-rc/latest/jenkins.war"
+            } else if (jenkins == "lts") {
+                jenkinsURl = mirror + "war-stable/latest/jenkins.war"
+            } else if (jenkins == "lts-rc") {
+                jenkinsURl = mirror + "war-stable-rc/latest/jenkins.war"
+            }
+
+            if (!isVersionNumber) {
+                echo 'Checking whether Jenkins WAR is available…'
+                sh "curl -ILf ${jenkinsURl}"
+            }
+            // Validation ended
+
+            // ATH
+            if (isLocalATH) { // Deal with already existing ATH sources
+                athSourcesFolder = athUrl - "file://"
+            } else {
+                dir(athSourcesFolder) {
+                    checkout changelog: true, poll: false, scm: [$class           : 'GitSCM', branches: [[name: athRevision]],
+                                                                 userRemoteConfigs: [[url: athUrl]]]
+                }
+            }
+            dir(athSourcesFolder) {
+                // We may need to run things in parallel later, so avoid several checkouts
+                stash name: "athSources"
```

Comment:
Not sure this is not early optimization. Why not just `checkout scm` wherever we need sources?

Reply:
Because potentially I can run in two nodes (one for categories and another for single tests) and in the future we may need to further parallelize, so I prefer to checkout in the linux node and stash and then unstash in all the parallel branches I may need. I also need sources to build the docker image and I want to do that only once

Reply:
Well not sure I buy that: `checkout scm` where need is simpler than `stash` and `unstash` at the right places. So again looks like a premature optimization.

I would tend to agree if this was to not rebuild artifacts. Sources should never matter. I suspect `stash`ing might even be slower than just `git clone`ing again. 
For what it's worth, just checked, a fresh `git clone` took less than a second on my machine.

Anyway.